diff --git a/lib/src/navigation_drawer.dart b/lib/src/navigation_drawer.dart
index 396aa11..3ad2ff7 100644
--- a/lib/src/navigation_drawer.dart
+++ b/lib/src/navigation_drawer.dart
@@ -73,8 +73,11 @@ class NavigationDrawer extends StatelessWidget {
     this.onDestinationSelected,
     this.selectedIndex = 0,
     this.tilePadding = const EdgeInsets.symmetric(horizontal: 12.0),
+    this.flex = 1,
   });
 
+  final int flex;
+
   /// The background color of the [Material] that holds the [NavigationDrawer]'s
   /// contents.
   ///
@@ -198,6 +201,7 @@ class NavigationDrawer extends StatelessWidget {
           children: <Widget>[
             ?header,
             Expanded(
+              flex: flex,
               child: Material(
                 type: MaterialType.transparency,
                 child: ListView(children: wrappedChildren),
@@ -416,7 +420,7 @@ class _NavigationDestinationBuilder extends StatelessWidget {
       child: _NavigationDestinationSemantics(
         child: SizedBox(
           height: navigationDrawerTheme.tileHeight ?? defaults.tileHeight,
-          child: inkWell,
+          child: Material(type: MaterialType.transparency, child: inkWell),
         ),
       ),
     );
