diff --git a/lib/src/bottom_sheet.dart b/lib/src/bottom_sheet.dart
index 52a8111..cdb5bfd 100644
--- a/lib/src/bottom_sheet.dart
+++ b/lib/src/bottom_sheet.dart
@@ -256,8 +256,11 @@ class BottomSheet extends StatefulWidget {
   }
 }
 
+typedef BottomSheetState = _BottomSheetState;
+
 class _BottomSheetState extends State<BottomSheet> {
   final GlobalKey _childKey = GlobalKey(debugLabel: 'BottomSheet child');
+  GlobalKey get childKey => _childKey;
 
   double get _childHeight {
     final renderBox = _childKey.currentContext!.findRenderObject()! as RenderBox;
@@ -268,14 +271,14 @@ class _BottomSheetState extends State<BottomSheet> {
 
   Set<WidgetState> dragHandleStates = <WidgetState>{};
 
-  void _handleDragStart(DragStartDetails details) {
+  void handleDragStart(DragStartDetails details) {
     setState(() {
       dragHandleStates.add(WidgetState.dragged);
     });
     widget.onDragStart?.call(details);
   }
 
-  void _handleDragUpdate(DragUpdateDetails details) {
+  void handleDragUpdate(DragUpdateDetails details) {
     assert(
       (widget.enableDrag || (widget.showDragHandle ?? false)) && widget.animationController != null,
       "'BottomSheet.animationController' cannot be null when 'BottomSheet.enableDrag' or 'BottomSheet.showDragHandle' is true. "
@@ -287,7 +290,7 @@ class _BottomSheetState extends State<BottomSheet> {
     widget.animationController!.value -= details.primaryDelta! / _childHeight;
   }
 
-  void _handleDragEnd(DragEndDetails details) {
+  void handleDragEnd(DragEndDetails details) {
     assert(
       (widget.enableDrag || (widget.showDragHandle ?? false)) && widget.animationController != null,
       "'BottomSheet.animationController' cannot be null when 'BottomSheet.enableDrag' or 'BottomSheet.showDragHandle' is true. "
@@ -378,9 +381,9 @@ class _BottomSheetState extends State<BottomSheet> {
       // no need to add it.
       if (!widget.enableDrag) {
         dragHandle = _BottomSheetGestureDetector(
-          onVerticalDragStart: _handleDragStart,
-          onVerticalDragUpdate: _handleDragUpdate,
-          onVerticalDragEnd: _handleDragEnd,
+          onVerticalDragStart: handleDragStart,
+          onVerticalDragUpdate: handleDragUpdate,
+          onVerticalDragEnd: handleDragEnd,
           child: dragHandle,
         );
       }
@@ -422,9 +425,9 @@ class _BottomSheetState extends State<BottomSheet> {
     return !widget.enableDrag
         ? bottomSheet
         : _BottomSheetGestureDetector(
-            onVerticalDragStart: _handleDragStart,
-            onVerticalDragUpdate: _handleDragUpdate,
-            onVerticalDragEnd: _handleDragEnd,
+            onVerticalDragStart: handleDragStart,
+            onVerticalDragUpdate: handleDragUpdate,
+            onVerticalDragEnd: handleDragEnd,
             child: bottomSheet,
           );
   }
diff --git a/lib/src/generated/bottom_sheet_defaults_m3.g.dart b/lib/src/generated/bottom_sheet_defaults_m3.g.dart
index 1abf23b..119e3c1 100644
--- a/lib/src/generated/bottom_sheet_defaults_m3.g.dart
+++ b/lib/src/generated/bottom_sheet_defaults_m3.g.dart
@@ -7,6 +7,8 @@
 //   packages/material_ui/tool/gen_defaults/bin/gen_defaults.dart.
 part of '../bottom_sheet.dart';
 
+typedef BottomSheetDefaultsM3 = _BottomSheetDefaultsM3;
+
 class _BottomSheetDefaultsM3 extends BottomSheetThemeData {
   _BottomSheetDefaultsM3(this.context)
     : super(
diff --git a/lib/src/scaffold.dart b/lib/src/scaffold.dart
index 168fc5b..c4ed4a1 100644
--- a/lib/src/scaffold.dart
+++ b/lib/src/scaffold.dart
@@ -3400,7 +3400,9 @@ class _DismissDrawerAction extends DismissAction {
 class ScaffoldFeatureController<T extends Widget, U> {
   const ScaffoldFeatureController._(this._widget, this._completer, this.close, this.setState);
   final T _widget;
+  T get widget => _widget;
   final Completer<U> _completer;
+  Completer<U> get completer => _completer;
 
   /// Completes when the feature controlled by this object is no longer visible.
   Future<U> get closed => _completer.future;
@@ -3412,6 +3414,9 @@ class ScaffoldFeatureController<T extends Widget, U> {
   final StateSetter? setState;
 }
 
+typedef StandardBottomSheet = _StandardBottomSheet;
+typedef StandardBottomSheetState = _StandardBottomSheetState;
+
 class _StandardBottomSheet extends StatefulWidget {
   const _StandardBottomSheet({
     super.key,
@@ -3477,12 +3482,12 @@ class _StandardBottomSheetState extends State<_StandardBottomSheet> {
     widget.onClosing?.call();
   }
 
-  void _handleDragStart(DragStartDetails details) {
+  void handleDragStart(DragStartDetails details) {
     // Allow the bottom sheet to track the user's finger accurately.
     animationCurve = Curves.linear;
   }
 
-  void _handleDragEnd(DragEndDetails details, {bool? isClosing}) {
+  void handleDragEnd(DragEndDetails details, {bool? isClosing}) {
     // Allow the bottom sheet to animate smoothly from its current position.
     animationCurve = Split(widget.animationController.value, endCurve: _standardBottomSheetCurve);
   }
@@ -3535,8 +3540,8 @@ class _StandardBottomSheetState extends State<_StandardBottomSheet> {
             animationController: widget.animationController,
             enableDrag: widget.enableDrag,
             showDragHandle: widget.showDragHandle,
-            onDragStart: _handleDragStart,
-            onDragEnd: _handleDragEnd,
+            onDragStart: handleDragStart,
+            onDragEnd: handleDragEnd,
             onClosing: widget.onClosing!,
             builder: widget.builder,
             backgroundColor: widget.backgroundColor,
@@ -3568,7 +3573,17 @@ class PersistentBottomSheetController
     this._isLocalHistoryEntry,
   ) : super._();
 
+  const PersistentBottomSheetController(
+    super.widget,
+    super.completer,
+    super.close,
+    StateSetter super.setState,
+    this._isLocalHistoryEntry,
+  ) : super._();
+
   final bool _isLocalHistoryEntry;
+
+  bool get isLocalHistoryEntry => _isLocalHistoryEntry;
 }
 
 class _ScaffoldScope extends InheritedWidget {
