diff --git a/lib/pages/common/slide/common_slide_page.dart b/lib/pages/common/slide/common_slide_page.dart
index 58ba68b7c..a6bb398ac 100644
--- a/lib/pages/common/slide/common_slide_page.dart
+++ b/lib/pages/common/slide/common_slide_page.dart
@@ -3,7 +3,8 @@ import 'dart:math' show max;
 
 import 'package:PiliPlus/common/widgets/gesture/horizontal_drag_gesture_recognizer.dart';
 import 'package:PiliPlus/utils/storage_pref.dart';
-import 'package:flutter/gestures.dart' show HorizontalDragGestureRecognizer;
+import 'package:flutter/gestures.dart'
+    show HorizontalDragGestureRecognizer, DeviceGestureSettings;
 import 'package:get/get.dart';
 import 'package:material_ui/material_ui.dart';
 
@@ -142,6 +143,10 @@ class SlideDragGestureRecognizer extends HorizontalDragGestureRecognizer {
   bool isPointerAllowed(PointerEvent event) {
     return isDxAllowed(event.localPosition.dx) && super.isPointerAllowed(event);
   }
+
+  @override
+  DeviceGestureSettings get gestureSettings =>
+      const DeviceGestureSettings(touchSlop: 5.0);
 }
 
 class TabBarDragGestureRecognizer
diff --git a/lib/common/widgets/scaffold/mini_scaffold.dart b/lib/common/widgets/scaffold/mini_scaffold.dart
index d1a0ca667..2b7760943 100644
--- a/lib/common/widgets/scaffold/mini_scaffold.dart
+++ b/lib/common/widgets/scaffold/mini_scaffold.dart
@@ -72,6 +72,7 @@ class MiniScaffoldState extends State<MiniScaffold>
     }
 
     final LocalHistoryEntry entry = LocalHistoryEntry(
+      popGestureEnabled: true,
       onRemove: () {
         if (!removedEntry &&
             _currentBottomSheet?.widget == bottomSheet &&
