diff --git a/packages/flutter/lib/src/rendering/paragraph.dart b/packages/flutter/lib/src/rendering/paragraph.dart
index 97a41dafb8d..7d807e1ad2c 100644
--- a/packages/flutter/lib/src/rendering/paragraph.dart
+++ b/packages/flutter/lib/src/rendering/paragraph.dart
@@ -385,6 +385,7 @@ class RenderParagraph extends RenderBox
   );
 
   final TextPainter _textPainter;
+  TextPainter get textPainter => _textPainter;
 
   // Currently, computing min/max intrinsic width/height will destroy state
   // inside the painter. Instead of calling _layout again to get back the correct
diff --git a/packages/flutter/lib/src/widgets/text.dart b/packages/flutter/lib/src/widgets/text.dart
index a3aa5fde862..51601f730af 100644
--- a/packages/flutter/lib/src/widgets/text.dart
+++ b/packages/flutter/lib/src/widgets/text.dart
@@ -1503,6 +1503,9 @@ typedef _SelectionInfo = ({int contentLength, SelectedContentRange? range});
 // must also be updated.
 // TODO(Renzo-Olivares): Remove after investigating a solution for overriding all
 // styles for children in an [InlineSpan] tree, see: https://github.com/flutter/flutter/issues/177952.
+
+typedef OverridingTextStyleTextSpanUtils = _OverridingTextStyleTextSpanUtils;
+
 class _OverridingTextStyleTextSpanUtils {
   static TextSpan applyTextSpacingOverrides({
     double? lineHeightScaleFactor,
